kernel & buildsys: change how internal GAP headers are included to minimize conflicts with OS headers

ptrrsn opened the pull request:
The header io.h from MinGW is hidden by gap's io.h which is causing compilation error. Changing -I to -iquote ensures that all gap header files are included with quotes, while the standard library header files are included with the <>.

This helps the work for #4157

dimpase commented:
OK, that's a good catch.

dimpase commented:
This is certainly the correct change, although 
one should check that this doesn't break building GAP packages (e.g. io) - they might need similar fixes, in fact.

ptrrsn commented:
Thank you all for the comments. I will be happy to make some improvements to unblock the native Windows build. For this PR, I can do the renaming, and I like @dimpase's idea to move all headers under a subdirectory to effectively namespace the header files and solve the conflicting names once and for all.

@fingolfin, forgive my very limited understanding, but were you suggesting hiding all the headers and exposing only gap_all.h instead? I am not familiar with the concept of kernel extensions. My understanding is that kernel extensions outside this repository use these headers too, and your suggestion involved a slow migration process, i.e., by maintaining backward compatibility for those external kernel extensions by still providing the rest of the .h headers for a specific period. Am I understanding your suggestion correctly? Since this sounds like something that can be done in parallel with the renaming, I will proceed with the subdirectory idea for now, and we can open a separate issue to discuss this long-term migration further if it is still needed, as I can see @dimpase has a different opinion on it (CMIIW). What do you think?

dimpase commented:
historically, GAP kernel modules were always built in-tree, that's why very little had been done as far as header namespaces go.

Many packages had to move to headers in subdirectories: to name a few in the area of algebra: gmp, flint/flintlib, cdd/cddlib. 

As far as distributions go, they would very much appreciate neat header handling, as dumping everything in `/usr/include` or in `/usr/local/include` isn't working at scale.

fingolfin commented:
@dimpase all GAP headers are already in a subdirectory `PREFIX/gap/` when installed, as is common for libraries. If you are suggesting to put all headers into a subdirectory in the source tree: that would be cumbersome for various reasons, and also simply not necessary. It may surprise you, but I actually thought about all these things years ago when I set out to get GAP ready to support `make install`. It's one of the various small things here and there that I simply did not yet get around. Luckily pretty much straightforward these days. 

Your suggestion to adding `-I libgap/` misses the entire point, that would reintroduce the problem with `io.h`.




@ptrrsn no, that's not my suggestion. Or rather: the GAP headers are already "hidden away" when GAP is installed, namely in `PREFIX/gap/`. However, this does not contribute to solving the `io.h` problem because at the same time the CPPFLAGS include `-IPREFIX/gap`.

What I am proposing is this instead:
1. we add a new file `src/extra/gap_all.h` to the repo which would basically consist of `#include "../gap_all.h"` plus a header comment
   - this makes use of the fact that the preprocessor first searches in the directory in which the current file is
   - `CPPFLAGS` then would only contain `-Ipath/to/gap.git/src/extra`
2. for `make install` we install this header like all others, i.e. as`PREFIX/gap/extra/gap_all.h`
   - `CPPFLAGS` then would only contain `-IPREFIX/gap/extra`
3. Packages should be changed to do `#include "gap_all.h"`
   - this was the plan all along, but I held off from doing it because some package authors don't like requiring "the latest" GAP
   - but this headers has been there there since GAP 4.12 from 2022, and requiring GAP >= 4.12 should be fine for all packages with kernel extensions (in fact I think we already did that last year, when we changed them to use `IsKernelExtensionAvailable` and `LoadKernelExtension` 
   - I will get right on that in fact, and make releases of most if not all those packages,

The main drawback is that it'll break packages with kernel extensions not in the GAP distro right now. But we can solve this for many of them by also adding a `src/extra/common.h` header which has identical content to `gap_all.h`.

ptrrsn commented:
Thank you everyone for the comments.

@fingolfin I updated my changes with my interpretation of your proposal. Could you check if my understanding is correct?

The way I tested it by:
1. running make and make install in Linux,
2. ./configure and make for tst/mockpkg,
3. write a C source code that include extra/gap_all.h and checked manually that the extra headers are installed in my Linux filesystem after make install,
4. The "io.h" problem in MinGW is resolved (by the iquote).

But looking at the CI breakage, it seems that those tests are not enough...

ptrrsn commented:
Ah, sorry, I accidentally triggered a CI run from a sync. I will try to fix the failures, except the "CI / testbuildsys; testmockpkg testinstall - NO_COVERAGE = 1 ..." seems to be an infrastructural failure. I also saw that in my other PR.

ptrrsn commented:
@fingolfin I think I have fixed all the CI problems (except the infrastructure failure one). As for HPC, I added all `src` instead of just `src/extra` in `SYSINFO_CPPFLAGS`. Without that special case, headers like `hpc/tls.h`, which has `#include "system.h", will not work because "system.h" is not in the same directory as "hpc/tls.h" and it is not in the `SYSINFO_CPPFLAGS`.

An alternative but more intrusive option is to change all such includes in `hpc` directories to be in the form of `#include "../system.h"`. I personally think this would be better than having a special case in Makefile.rules because it avoids different expectations between hpc vs non-hpc. If this one is preferred, I can make this change in a separate PR so as not to clutter the current PR further. What do you think?

fingolfin commented on src/extra/compiled.h:
```diff
@@ -0,0 +1,19 @@
+**  included by GAP kernel extensions.
+*/
+
+#ifndef SRC_EXTRA_COMPILED_H
+#define SRC_EXTRA_COMPILED_H
```
```suggestion
#ifndef GAP_EXTRA_COMPILED_H
#define GAP_EXTRA_COMPILED_H
```

fingolfin commented on src/extra/compiled.h:
```diff
@@ -0,0 +1,19 @@
+
+#include "../compiled.h"
+
+#endif // SRC_EXTRA_COMPILED_H
```
```suggestion
#endif // GAP_EXTRA_COMPILED_H
```

fingolfin commented on src/extra/gap_all.h:
```diff
@@ -0,0 +1,19 @@
+**  included by GAP kernel extensions.
+*/
+
+#ifndef SRC_EXTRA_GAP_ALL_H
+#define SRC_EXTRA_GAP_ALL_H
+
+#include "../gap_all.h"
+
+#endif // SRC_EXTRA_GAP_ALL_H
```
```suggestion
#ifndef GAP_EXTRA_GAP_ALL_H
#define GAP_EXTRA_GAP_ALL_H

#include "../gap_all.h"

#endif // GAP_EXTRA_GAP_ALL_H
```

fingolfin commented on Makefile.rules:
```diff
@@ -193,12 +193,12 @@ GAP_CPPFLAGS += -I$(builddir)/build
 # The other source files are relative to the srcdir
-GAP_CPPFLAGS += -I$(srcdir)/src
-SYSINFO_CPPFLAGS += -I$(abs_srcdir)/src
+GAP_CPPFLAGS += -iquote$(srcdir)/src
```
Why `-iquote` ?

fingolfin replied:
OK, fair enough. Can you please add that as a comment explaining this rationale, so that just in case someone has trouble with this line in the future they know why it  uses `-iquote` instead of `-I` ?

fingolfin commented on tst/mockpkg/src/mockpkg.c:
```diff
@@ -1,4 +1,4 @@
-#include <gap_all.h>    // GAP headers
+#include "gap_all.h"    // GAP headers
```
But the work you are doing here is meant to allow this to be used with `<...>` so I would suggest we should test that here...
```suggestion
#include <gap_all.h>    // GAP headers
```

fingolfin commented on tst/testkernel/dstruct.c:
```diff
@@ -2,11 +2,12 @@
 #include <stdlib.h>
 #include <string.h>
 
+#include "libgap-api.h"
```
Hmm, maybe `libgap-api.h` should get the same treatment as `gap_all.h`, just intended for users of the `libgap` instead of kernel extensions...

fingolfin commented on Makefile.rules:
```diff
@@ -193,12 +193,12 @@ GAP_CPPFLAGS += -I$(builddir)/build
-# For backwards compatibility with certain packages and their kernel
-# extensions, also add the root of the srcdir to the package CPPFLAGS
-SYSINFO_CPPFLAGS += -I$(abs_srcdir)
+ifeq ($(HPCGAP),yes)
+SYSINFO_CPPFLAGS += -I$(abs_srcdir)/src
+endif
```
For most purposes HPC-GAP is essentially dead, so I am fine with either this workaround, or, as you suggested, the files in there using relative paths in their `#include`.

Perhaps a compromise would be to only adapt the *header files* in `src/hpc` to use relative paths. That would only affect 12 files and in each just 1-3 lines

ptrrsn replied:
Changed the header files in src/hpc to use relative paths and removed this extra condition.

ptrrsn replied:
ah sorry, I have just updated the PR. Re-resolving this back.

ptrrsn commented on src/extra/common.h:
```diff
@@ -0,0 +1,19 @@
+**
+**  SPDX-License-Identifier: GPL-2.0-or-later
+**
+**  This header includes most of the other GAP headers, and is meant to be
+**  included by GAP kernel extensions.
```
Ah, nevermind, using `#include "gap/common.h"` made everything very messy because in the src directory, `ffdata.h` and `common.h` belong to different directories (`build` vs `src`) while during make install, they belong to the same directory. Because `ffdata.h` and `version.h` only depends on `common.h`, I instead add a rule to copy the `common.h` to build during `make`.

ptrrsn commented:
Sorry, I resolved them on my local branch and was preparing to push, but I found some errors. I will push first before replying to/resolving your comments next time to avoid such confusion.

I have addressed your comments in this last update.

ptrrsn commented on src/extra/compiled.h:
```diff
@@ -0,0 +1,19 @@
+**  SPDX-License-Identifier: GPL-2.0-or-later
+**
+**  This header includes most of the other GAP headers, and is meant to be
+**  included by GAP kernel extensions.
```
re-resolving after local push.

ptrrsn commented on src/extra/compiled.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef SRC_EXTRA_COMPILED_H
+#define SRC_EXTRA_COMPILED_H
+
+#include "../compiled.h"
```
This one actually caused problem during test-compile, because `and_filter.g.dynamic.c` needs `GVar` definition from `compiled.h`, so I am reverting back to `../compiled.h`.

ptrrsn commented:
The PR is ready for review. Sorry for the confusion!

fingolfin commented on src/extra/libgap-api.h:
```diff
@@ -0,0 +1,17 @@
+
+#include "../libgap-api.h"
+
+#endif // GAP_EXTRA_LIBGAP_API_H
```
```suggestion
#endif // GAP_EXTRA_LIBGAP_API_H

```

ptrrsn commented on Makefile.rules:
```diff
@@ -548,6 +546,14 @@ ffgen: $(srcdir)/etc/ffgen.c src/common.h build/config.h
+# Copy common.h to the build directory to be used by ffdata
+########################################################################
+
+build/common.h: $(srcdir)/src/common.h
```
Without this, the following will fail:
```
./autogen.sh && ./configure && make && cd pkg/io && make
```
with something like:
```
   GAC     src/io.c => gen/src/io.o
In file included from /gap_root/gap/src/extra/../finfield.h:59,
                 from /gap_root/gap/src/extra/../gap_all.h:35,
                 from /gap_root/gap/src/extra/gap_all.h:17,
                 from src/io.c:14:
/gap_root/gap/build/ffdata.h:6:10: fatal error: common.h: No such file or directory
    6 | #include "common.h"
      |          ^~~~~~~~~~
compilation terminated.
make: *** [Makefile.gappkg:119: gen/src/io.o] Error 1
```

Because `ffdata.h` is generated in the `build` directory, it could not find `common.h`, which is located either at `build/gap` or at `src` directory.

But we cannot simply change that to `#include "gap/common.h"` because `make install` will put `ffdata.h` and `common.h` (and the rest of gap headers) in the same directory. So, the simplest solution here that I can think of is to just copy common.h to `build` directory.

fingolfin replied:
Actually, instead of messing about with `common.h`, it should also work to drop the `#include "common.h"` from `ffdata.h`, and then adding `#include "gap_all.h"` to the start of `ffdata.c` (both changes would have to be made to `etc/ffgen.c`).

I don't see this as fundamentally better or worse; the one thing I like about it is that it is one less "moving part" in the build system and one less difference between installed and not-installed GAP. Alas, I don't feel strongly about which we pick right now.

(Perhaps you or @ChrisJefferson hav some opinions?)

ptrrsn replied:
I added symlink.

fingolfin commented on src/extra/libgap-api.h:
```diff
@@ -0,0 +1,17 @@
+**  to list here. Please refer to the COPYRIGHT file for details.
+**
+**  SPDX-License-Identifier: GPL-2.0-or-later
+**  This header includes LibGAP API, the API for using GAP as shared library.
```
```suggestion
**
**  This header includes LibGAP API, the API for using GAP as shared library.
```

fingolfin commented on Makefile.rules:
```diff
@@ -192,13 +192,11 @@ SYSINFO_CPPFLAGS =
-# For backwards compatibility with certain packages and their kernel
-# extensions, also add the root of the srcdir to the package CPPFLAGS
-SYSINFO_CPPFLAGS += -I$(abs_srcdir)
+# The other source files are relative to the srcdir. The iquote is used
+# instead of -I to avoid conflicting io.h between MinGW standard library
+# and GAP's.
+GAP_CPPFLAGS += -iquote$(srcdir)/src -I$(srcdir)/src/extra
```
Sorry, I think I asked it before but I can't find it right now among the various resolved comments, so I'll just ask again (?): why do we need the `-iquote` again? I don't mean "instead of `-I`, I mean: "at all". I.e. why doesn't this work:
```suggestion
# The other source files are relative to the srcdir. The iquote is used
# instead of -I to avoid conflicting io.h between MinGW standard library
# and GAP's.
GAP_CPPFLAGS += -I$(srcdir)/src/extra
```
I would have thought it is for `common.h` being includable from `ffdata.h` but for that you are already adding `build/common.h` ?

ptrrsn replied:
Hmm, there are several reasons why it is needed, but I think mostly because some of the GAP headers are used from the tst directory. To summarize some that benefits from iquote:
1. we need the iquote here is because the `common.h` included from `ffdata.h` has `#include "debug.h"` in it. But this one can be resolved alongside your other comment about the alternative to copying `common.h` and applying it to `debug.h`. I will reply to your other comment separately there.
2. The `tst/testlibgap/api.c` and `tst/testlibgap/trycatch.c` contains `#include "intobj.h"` and `#include "trycatch.h"` respectively. But both can be replaced by `gap_all.h`.
3. There are includes to "baltree.h" or "dynarray.h" that are not part of the headers in `extra` from for example `testkernel/dstruct.c`.

So I think keeping the `iquote` will be cleaner and nicer for the test directory, and we can keep all those direct includes from tests.

fingolfin replied:
But those kernel tests are something very special and could receive special treatment, while `-iquote` has portability issues. Regarding the specific points
1. we could simply inline `debug.h` into `common.h`
2. those `tst/testlibgap` headers could indeed use `gap_all.h` for now
   - ... though regarding `intobj.h` being used in `api.c` it would be better to apply the patch below
```patch
diff --git a/tst/testlibgap/api.c b/tst/testlibgap/api.c
index ab232cf4ca..fbfcf73a97 100644
--- a/tst/testlibgap/api.c
+++ b/tst/testlibgap/api.c
@@ -2,7 +2,7 @@
  * Small program to test libgap api functions.
  *
  * Note that we only test whether the API functions
- * work to a reasonable extent, the functionality of 
+ * work to a reasonable extent, the functionality of
  * the GAP system is tested in extensive tests of the
  * system itself.
  *
@@ -11,7 +11,6 @@
  */
 
 #include "common.h"
-#include "intobj.h"
 
 #include <string.h>
 
@@ -100,7 +99,7 @@ void matrices(void)
     Obj mat, val, row, ret;
 
     mat = GAP_NewPlist(1);
-    val = INTOBJ_INT(42);
+    val = GAP_NewObjIntFromInt(42);
 
     assert(!GAP_IsMatrixOrMatrixObj(mat));   // empty list, not yet a matrix
     assert(!GAP_IsMatrixObj(mat));
@@ -113,8 +112,8 @@ void matrices(void)
     assert(!GAP_IsMatrix(val));
 
     row = GAP_NewPlist(2);
-    GAP_AssList(row, 1, INTOBJ_INT(1));
-    GAP_AssList(row, 2, INTOBJ_INT(2));
+    GAP_AssList(row, 1, GAP_NewObjIntFromInt(1));
+    GAP_AssList(row, 2, GAP_NewObjIntFromInt(2));
     GAP_AssList(mat, 1, row);
     assert(!GAP_IsMatrixOrMatrixObj(row));
     assert(!GAP_IsMatrixObj(row));
```
3. finally `testkernel/dstruct.c` could probably be changed to use `#include "gap/baltree.h"` and `#include "gap/dynarray.h"` -- or alternatively, it could be built with special CPPFLAGS including `-Isrc`. After all, nobody needs to run this test under MingGW

ptrrsn replied:
Thank you for the patch! I used your change and it seems to works nicely! I removed iquote and did a few more things to support it, including:
- a bunch more `hpc` includes changes to relative path
- `rename /build/main.c` to `/build/main_modified.c` during `make install`.

ptrrsn commented:
> I'm at a conference and don't have access to my usual array of weird machines, but I am skeptical that `-iquote` is portable. It looks like clang does support it, even though it's not documented. In any case, POSIX only has `-I`: https://pubs.opengroup.org/onlinepubs/9799919799/utilities/c17.html

Both clang (https://clang.llvm.org/docs/ClangCommandLineReference.html) and gcc (https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html) seem to support `iquote` and they are both documented. If those are insufficient, I can try to remove the `iquote` use here.

ptrrsn commented:
Thank you for the reviews! I have updated the PR and addressed all the comments.

ptrrsn commented on Makefile.rules:
```diff
@@ -514,12 +510,12 @@ gap$(EXEEXT): build/obj/src/main.c.o $(OBJS) cnf/GAP-LDFLAGS cnf/GAP-LIBS cnf/GA
 
 # generate a modified copy of main.c for use by the `gap-install` binary
-build/main.c: src/main.c config.status
+build/gap/main_modified.c: src/main.c config.status
```
Ah nevermind, the failing CI shows that we indeed need to rename the `main.c`. I renamed it back to `main_modified.c`

fingolfin replied:
Well `build/gap` is just a symlink to `src`, which is why `build/gap/main.c` won't work: it would be overwriting `src/main.c`. 

But `build/gap/main_modified.c` likewise is problematic because it would place a generated file into `src` which the build system is not supposed to do ...

Some alternative ideas:
1. we could modify `main.c` to use `gap_all.h` to replace most of `#include` (we still need `config.h`, but that one is already in `build` so no problem
2. we could replace the entire "patch `main.c`"  approach by a special build rule which passes a suitable `-D name=value` option to the C compiler to define `SYS_DEFAULT_PATHS` (this is in POSIX). IIRC the main reason I didn't use that approach back when is that the idea of figuring out the correct quoting of everything to pass from Makefile to shell to C seemed like a sure path to madness :joy: alas you might be a braver soul than me
3. one could of course also symlink all "missing" headers for `main.c` into `build` ... 
4. ... ?

ptrrsn replied:
How about creating `src/system/system_all.h` that redirects all the system files (like `src/extra/gap_all.h`), and add it to `GAP_CPPFLAGS` with `-I`? Then, I can revert this change and let main.c to `#include "system_all.h"` and `#include "gap_all.h"` instead.

ptrrsn replied:
I feel that it will be conceptually simpler and cleaner because it uses the same pattern as `src/extra` and less invasive to the `build` structure.

ptrrsn replied:
I have added the `system` directory in the last update.

fingolfin commented:
So... please don't hate me for asking this, considering how much effort you have put into this PR by now, but, just thinking about this again now, remind me: why do we (resp. you, thank you!) go to all this effort and ever growing list of complications, instead of simply renaming `io.h` to `gap_io.h` or `inout.h` or `input_output.h` or whatever... ?

ptrrsn commented:
I am trying to resolve the problem cleanly by implementing your long-term suggestion, which, in my understanding, is preferred for GAP. If possible, I would prefer not to introduce technical debt by renaming "io.h" as a short-term fix. The long-term suggestion that I meant can be found in your comment here: https://github.com/gap-system/gap/pull/6007#issuecomment-2987070411, which also seems to provide a better boundary between GAP and its packages (so not just fixing the "io.h" conflicts problem).

However, I can also revert this and perform a simple renaming instead if that's what you prefer. Please let me know your preference :-)

fingolfin commented:
I am honestly on the fence. Complications in the build system also are a form of technical debt.

I think the idea of adding the `extra` subdir and letting packages by default only use that, is a good one.

But a lot of the other hoops you had to jump through all revolve around letting GAP source files outside of `src` (such as `ffdata.c`, a modified `main.c`) access the header files and this requires more and more contortions.

Like, it is OK to inline `debug.h` into `common.h` but we really only did it as a workaround. 


Then again, overall the PR is not too bad as it is right now. I am inclined to just merge it. But I'd love to hear what e.g. @ChrisJefferson thinks

ptrrsn commented:
No worries, I will follow whatever you decide, as you are the one who has deep insight into what will be best for the repository :-)

But I also agree that the current change makes the build system more complicated. How about we do something in the middle:
- We expose only `extra` directory on SYSINFO_CPPFLAGS (i.e., the last state of the PR)
- For GAP_CPPFLAGS, we rename `io.h` to `input_output.h` and keep the original `-I$(srcdir)/src` instead of `-I$(srcdir)/src/extra -I$(srcdir)/src/system`.

That way, we still restrict packages to only be able to include the `extra` directory while maintaining the simplicity for the internal GAP code (then, we can revert all the complications on `build` directory: the inlining `debug.h`, etc) + solving the "io.h" problem. This may be a better option, what do you think? @fingolfin

fingolfin commented:
So @ChrisJefferson kinda convinced me, but then I also had another think, and realized we might be able to simplify a few things again...

So I tweaked this PR a bit to avoid the need for `build/common.h` and `src/system/system_all.h`, which then allowed me to restored `src/debug.h`. Please have a look at the commits I added.

I'd be happy to merge this now if you are also OK with it, @ptrrsn

ptrrsn commented:
Wow, thank you for the additional patches! They look really great!
